File: rb/lib/selenium/webdriver/bidi/serialization/record.rb
# frozen_string_literal: true

# Licensed to the Software Freedom Conservancy (SFC) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The SFC licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

module Selenium
  module WebDriver
    class BiDi
      module Serialization
        # Immutable value type for the generated protocol classes. +Record.define(spec)+
        # bakes each field's wire facts and returns a +::Data+ subclass with serialization.
        #
        #   Cookie = Record.define(name: 'name', value: {wire_key: 'value', ref: 'Network::BytesValue'})
        #
        # @api private
        class Record < ::Data
          # Named Field, not Member, to avoid colliding with +::Data#members+.
          Field = ::Data.define(:name, :wire_key, :nullable, :ref, :list, :fixed, :enum, :required, :primitive,
                                :scalar, :const)

          def self.define(**spec)
            extensible = spec.delete(:extensible) || false
            fields = spec.map { |name, meta| field(name, meta) }
            names = fields.map(&:name)
            names << :extensions if extensible

            klass = super(*names)
            fields.freeze
            # Singleton methods are inherited by `X = Record.define(…)`; ivars would not.
            klass.define_singleton_method(:fields) { fields }
            klass.define_singleton_method(:extensible?) { extensible }
            klass.include(Serializable)
            # Capture ::Data's generated +new+, then prepend (not include) Deserializer so
            # its +new+ overrides it — outbound +new+ adds validation, while inbound
            # +from_json+ builds directly via the captured constructor. Bound to +self+ so a
            # subclass builds itself, not the base.
            data_new = klass.singleton_class.instance_method(:new)
            klass.singleton_class.prepend(Deserializer)
            klass.define_singleton_method(:construct) { |**attributes| data_new.bind_call(self, **attributes) }
            klass.singleton_class.send(:private, :construct)
            klass
          end

          def self.field(name, meta)
            meta = {wire_key: meta} if meta.is_a?(::String)
            Field.new(name: name.to_sym, wire_key: meta.fetch(:wire_key, name.to_s),
                      nullable: meta[:nullable] || false, ref: meta[:ref],
                      list: meta[:list] || false, fixed: meta.fetch(:fixed, UNSET), enum: meta[:enum],
                      required: meta.fetch(:required, true), primitive: meta[:primitive],
                      scalar: meta[:scalar], const: meta.fetch(:const, UNSET))
          end
          private_class_method :field

          # Inbound construction: the keyword +new+ (validated) and the wire +from_json+.
          #
          # @api private
          module Deserializer
            def new(**kwargs)
              # Start from what was passed so ::Data's constructor rejects an unknown key, then fill
              # each field with its value or UNSET (omitted), forcing fixed discriminators.
              attributes = kwargs.dup
              fields.each { |f| attributes[f.name] = fixed?(f) ? f.fixed : attributes.fetch(f.name, UNSET) }
              attributes[:extensions] = kwargs.fetch(:extensions, {}) if extensible?
              validate_values(attributes)
              construct(**attributes)
            end

            # Inbound: builds from the wire. A missing required field raises (in +wire_value+); enum
            # tokens are mapped back to symbols and an unrecognized one raises (in +read+); an
            # undeclared property is captured silently (extensible) or warned and dropped (closed)
            # — strict on shape, lenient on extras.
            def from_json(json_payload)
              unless json_payload.is_a?(::Hash)
                raise Error::SerializationError, "#{name} expected an object on the wire, got #{json_payload.inspect}"
              end

              attributes = fields.to_h do |f|
                [f.name, wire_value(f, json_payload)]
              end
              undeclared = extra(json_payload)
              if extensible?
                attributes[:extensions] = undeclared # the spec sanctions these extras; preserve them silently
              else
                warn_undeclared(undeclared) unless undeclared.empty?
              end
              construct(**attributes)
            end

            private

            # Checks each field's value: a required field cannot be omitted (UNSET), a non-nullable
            # field cannot be nil (nil is neither a value nor the UNSET omit-sentinel, so it would be
            # silently dropped on the wire), a nullable-const field must carry its literal (not some
            # other value), a primitive field must be the matching Ruby type, and an enum field must be
            # in its allowed set. The enum constant is resolved lazily so a cross-domain enum need not be
            # loaded first. Outbound only (from +new+); inbound presence/primitive/enum are checked
            # separately in +wire_value+/+read+.
            def validate_values(attributes)
              fields.each do |f|
                value = attributes[f.name]
                raise ::ArgumentError, "#{name}##{f.name} is required" if UNSET.equal?(value) && f.required
                raise ::ArgumentError, "#{name}##{f.name} cannot be nil" if value.nil? && !f.nullable
                next if value.nil? || UNSET.equal?(value)

                validate_present(f, value)
              end
            end

            # Checks a field that carries an actual value (neither omitted nor nil): a nullable-const
            # field against its literal, list/scalar shape, primitive type (lists excepted, as inbound
            # does), ref type, and enum membership (resolved lazily so a cross-domain enum need not load first).
            def validate_present(field, value)
              validate_const(field, value)
              check_outbound_shape(field, value)
              check_outbound_primitive(field, value) unless field.list
              validate_ref(field, value) if field.ref
              Serialization.validate!("#{name}##{field.name}", value, Protocol.const_get(field.enum)) if field.enum
            end

            # Outbound mirror of read_ref: a ref-typed value must be the type it declares, so a wrong
            # record or a value no union variant accepts is a caller error caught here, not a browser
            # round-trip. Shape is already checked, so a list is an Array.
            def validate_ref(field, value)
              klass = (@refs ||= {})[field.name] ||= Protocol.const_get(field.ref)
              field.list ? validate_ref_list(field, klass, value) : validate_ref_value(field, klass, value)
            end

            # Mirrors read_list: a scalar field is a [key, value] map, a nested list recurses, otherwise
            # each element is checked against the ref.
            def validate_ref_list(field, klass, list)
              list.each do |element|
                if field.scalar
                  validate_ref_entry(field, klass, element)
                elsif element.is_a?(::Array)
                  validate_ref_list(field, klass, element)
                else
                  validate_ref_value(field, klass, element)
                end
              end
            end

            # A [key, value] map entry: the key may be a variant or a bare scalar, the value is a variant.
            def validate_ref_entry(field, klass, element)
              unless element.is_a?(::Array) && element.size == 2
                raise ::ArgumentError, "#{name}##{field.name} expected a [key, value] pair, got #{element.inspect}"
              end

              key, value = element
              key.is_a?(Serializable) ? validate_ref_value(field, klass, key) : check_outbound_scalar(field, key)
              validate_ref_value(field, klass, value)
            end

            # A record ref must be an instance of that record; a union ref must be one the union accepts.
            def validate_ref_value(field, klass, value)
              return if klass < Union ? klass.valid_outbound?(value) : value.is_a?(klass)

              raise ::ArgumentError, "#{name}##{field.name} expected #{field.ref}, got #{value.inspect}"
            end

            # Outbound mirror of scalar_value: a bare map key must match one of the arm's primitives.
            def check_outbound_scalar(field, value)
              checks = Array(field.scalar).filter_map { |primitive| PRIMITIVE_CHECKS[primitive] }
              return if checks.empty? || checks.any? { |check| check.call(value) }

              raise ::ArgumentError,
                    "#{name}##{field.name} expected #{Array(field.scalar).join(' or ')}, got #{value.inspect}"
            end

            # A nullable constant (`literal / null`) is caller-settable but its only non-null value is
            # the literal, so a value that is neither the literal nor nil (nil is handled above) is a
            # local error rather than a wire round-trip. A non-const field carries UNSET here and passes.
            def validate_const(field, value)
              return if UNSET.equal?(field.const) || value == field.const

              raise ::ArgumentError, "#{name}##{field.name} must be #{field.const.inspect}, got #{value.inspect}"
            end

            # Outbound mirror of check_shape: a list-typed arg must be an array, a scalar-shaped one
            # (enum or ref, not a list) must not — a local ArgumentError, not a wire round-trip.
            def check_outbound_shape(field, value)
              return if field.list == value.is_a?(::Array)
              return unless field.list || field.enum || field.ref

              kind = field.list ? 'a list' : 'a single value'
              raise ::ArgumentError, "#{name}##{field.name} expected #{kind}, got #{value.inspect}"
            end

            # Outbound mirror of check_primitive: a primitive-typed arg (`string`/`integer`/…) must be
            # the matching Ruby type, so a caller mistake (a string width, a float count) is a local
            # ArgumentError here rather than a rejection the browser reports a round-trip later. A field
            # with no primitive descriptor (enum, ref, opaque) passes; lists are skipped, as inbound does.
            def check_outbound_primitive(field, value)
              check = PRIMITIVE_CHECKS[field.primitive]
              return if check.nil? || check.call(value)

              raise ::ArgumentError, "#{name}##{field.name} expected #{field.primitive}, got #{value.inspect}"
            end

            def fixed?(field)
              !UNSET.equal?(field.fixed)
            end

            # A required field absent from the response cannot yield a valid typed object, so it raises
            # rather than substitute a placeholder or represent the field as omitted; a remote end that
            # lags the schema is handled by a project schema override, not by runtime tolerance.
            def wire_value(field, json_payload)
              return field.fixed if fixed?(field)
              return read(field, json_payload[field.wire_key]) if json_payload.key?(field.wire_key)
              return UNSET unless field.required

              raise Error::SerializationError, "#{name}##{field.name} is required but was missing from the response"
            end

            def read(field, raw)
              if raw.nil?
                return raw if field.nullable

                raise Error::SerializationError, "#{name}##{field.name} received null but is not nullable"
              end
              check_shape(field, raw)
              return Serialization.to_symbol("#{name}##{field.name}", raw, enum_hash(field)) if field.enum

              if field.ref.nil?
                return raw if field.list

                check_primitive(field, raw)
                # A whole number is exact in both types, so the declared type is held with nothing lost.
                return field.primitive == 'integer' && raw.is_a?(::Float) ? raw.to_i : raw
              end

              read_ref(field, raw)
            end

            # Reads a ref-typed value into its class. A `scalar` position is an inline union with
            # a scalar arm collapsed onto its union ref (a map's string keys): a non-object leaf
            # passes through instead of being handed to the object_only union, but only when it
            # matches the arm's primitive (+scalar+ carries it). A list recurses per element.
            def read_ref(field, raw)
              klass = (@refs ||= {})[field.name] ||= Protocol.const_get(field.ref)
              return read_list(field, raw, klass) if field.list

              field.scalar && !raw.is_a?(::Hash) ? scalar_value(field, raw) : klass.from_json(raw)
            end

            # A declared list must arrive as an array; a scalar-shaped field (enum or ref, not a
            # list) must not. An opaque field carries no shape descriptor, so it passes through.
            def check_shape(field, raw)
              return if field.list == raw.is_a?(::Array)
              return unless field.list || field.enum || field.ref

              raise Error::SerializationError,
                    "#{name}##{field.name} expected #{field.list ? 'a list' : 'a single value'}, got #{raw.inspect}"
            end

            # The check a schema primitive admits, by JSON kind rather than Ruby class: `number` is
            # any Numeric (JSON has one number type), and `integer` is any whole one — a browser is
            # free to send `5` or `5.0` (JS has no int/float split), while a fractional value like
            # 1.5 is a real mismatch. A field with no primitive descriptor is left unchecked.
            WHOLE_FLOAT = ->(value) { value.is_a?(::Float) && value.finite? && (value % 1).zero? }
            PRIMITIVE_CHECKS = {
              'string' => ->(value) { value.is_a?(::String) },
              'boolean' => ->(value) { value.is_a?(::TrueClass) || value.is_a?(::FalseClass) },
              'number' => ->(value) { value.is_a?(::Numeric) },
              'integer' => ->(value) { value.is_a?(::Integer) || WHOLE_FLOAT.call(value) }
            }.freeze

            def check_primitive(field, raw)
              check = PRIMITIVE_CHECKS[field.primitive]
              return if check.nil? || check.call(raw)

              raise Error::SerializationError, "#{name}##{field.name} expected #{field.primitive}, got #{raw.inspect}"
            end

            def enum_hash(field)
              (@enums ||= {})[field.name] ||= Protocol.const_get(field.enum)
            end

            # Parses each element. A `scalar` field is a map encoded as `[key, value]` pairs, so
            # every element must be a 2-item pair — each is read as one, and a malformed entry is
            # rejected. Non-scalar lists recurse into nested lists; other elements deserialize.
            def read_list(field, raw, klass)
              raw.map do |element|
                if field.scalar
                  read_map_entry(field, element, klass)
                elsif element.is_a?(::Array)
                  read_list(field, element, klass)
                else
                  klass.from_json(element)
                end
              end
            end

            # A map entry is a `[key, value]` pair. The key is `Ref / text` — an object key
            # deserializes, a bare-string key passes through once validated against the arm's
            # primitive. The value is the object-only Ref and always deserializes, so a bare scalar
            # there is rejected (object_only holds at the value position). A non-pair element is a
            # malformed entry and is rejected outright.
            def read_map_entry(field, element, klass)
              unless element.is_a?(::Array) && element.size == 2
                raise Error::SerializationError,
                      "#{name}##{field.name} expected a [key, value] pair, got #{element.inspect}"
              end

              key, value = element
              key = key.is_a?(::Hash) ? klass.from_json(key) : scalar_value(field, key)
              [key, klass.from_json(value)]
            end

            # A bare scalar at a scalar-tolerant union position must match one of the union's
            # scalar-arm primitives (+scalar+ is a primitive name or an array of them); a
            # wrong-typed scalar (a number where a string is expected) is a wire error, not
            # something to pass through. An unrecognized primitive (none in PRIMITIVE_CHECKS) is
            # left unchecked, matching the lenient default elsewhere.
            def scalar_value(field, value)
              checks = Array(field.scalar).filter_map { |primitive| PRIMITIVE_CHECKS[primitive] }
              return value if checks.empty? || checks.any? { |check| check.call(value) }

              raise Error::SerializationError,
                    "#{name}##{field.name} expected #{Array(field.scalar).join(' or ')}, got #{value.inspect}"
            end

            def extra(json_payload)
              known = (@wire_keys ||= fields.map(&:wire_key))
              json_payload.except(*known)
            end

            # Forward-compat signal: a property a closed type does not model is dropped and warned so
            # schema drift is visible (an extensible type keeps its extras silently — the spec sanctions
            # them). Tagged +:bidi_undeclared_property+ so a caller can silence it via +logger.ignore+.
            def warn_undeclared(undeclared)
              undeclared.each_key do |key|
                WebDriver.logger.warn("#{name} received an undeclared property: #{key.inspect}",
                                      id: :bidi_undeclared_property)
              end
            end
          end

          # @api private
          module Serializable
            def self.as_json(value)
              case value
              when Serializable then value.as_json
              when ::Array then value.map { |element| as_json(element) }
              when ::Hash then value.transform_values { |element| as_json(element) }
              else value
              end
            end

            # Omit UNSET fields; emit null only for nullable ones.
            def as_json(*)
              payload = {}
              self.class.fields.each do |f|
                value = public_send(f.name)
                next if UNSET.equal?(value)
                next if value.nil? && !f.nullable

                value = Serialization.to_wire(value, Protocol.const_get(f.enum)) if f.enum
                payload[f.wire_key] = Serializable.as_json(value)
              end
              merge_extensions!(payload) if self.class.extensible? && !extensions.empty?
              payload
            end

            private

            # Merge the passthrough extras onto the wire, erroring rather than letting an extra whose key
            # is a declared field's wire key silently clobber that typed value; an extra is by definition
            # a field the spec does not declare. Keys are stringified first so a symbol key (e.g. `name:`)
            # cannot slip past the guard and then reappear as a duplicate wire key once serialized. The
            # single gate every outbound path funnels through: +new+, +with+, and in-place mutation.
            def merge_extensions!(payload)
              extras = extensions.transform_keys(&:to_s)
              collisions = extras.keys & self.class.fields.map(&:wire_key)
              unless collisions.empty?
                raise ::ArgumentError, "#{self.class.name} extensions shadow declared fields: #{collisions.join(', ')}"
              end

              payload.merge!(extras)
            end
          end
        end
      end
    end # BiDi
  end # WebDriver
end # Selenium

File: javascript/selenium-webdriver/bidi/serialization/record.js
// Licensed to the Software Freedom Conservancy (SFC) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The SFC licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

const { register, resolve } = require('./registry')

class ValidationError extends Error {}

// Validates a *present* value against a resolved type node, and returns the value to
// assign for it — usually the same value, but for a nested ref-to-record/union parsed
// inbound (fromWire()/build() results, not the raw wire object) and for list/map/inline
// record values (a fresh, deeply-frozen copy, so validity can't be corrupted after the
// fact by mutating an array or object a caller still holds a reference to).
// Outbound nested refs deliberately keep the caller's own value instead of the
// newly-constructed instance — the nested `new referenced.RecordClass(value)` /
// `referenced.build(value)` call below exists only to validate, matching the
// constructor's existing outbound behavior of trusting the caller's own object shape.
// `direction` only affects how a nested ref-to-record/union is itself validated/parsed.
function validateValue(typeNode, value, path, direction) {
  if (value === null) {
    // `primitive: 'null'` (project_bidi_schema.mjs's projectRef(), for a type whose
    // every alternative was null) means null itself is the valid value — accept it
    // even though `nullable` (a *different* value also being permitted alongside a
    // non-null base type) wasn't separately set.
    if (typeNode.nullable || typeNode.primitive === 'null') return null
    throw new ValidationError(`${path}: null is not allowed`)
  }

  if (typeNode.primitive !== undefined) {
    // A present, non-null value never satisfies a `primitive: 'null'` node — value
    // being null is the only thing the branch above would have already returned for.
    if (typeNode.primitive === 'null') {
      throw new ValidationError(`${path}: expected null, got ${typeof value}`)
    }
    const expected = { string: 'string', integer: 'number', number: 'number', boolean: 'boolean' }[typeNode.primitive]
    if (expected && typeof value !== expected) {
      throw new ValidationError(`${path}: expected ${typeNode.primitive}, got ${typeof value}`)
    }
    // JSON has no representation for NaN/±Infinity — reject them for both numeric
    // primitives before the integer-specific check narrows further. (Number.isInteger
    // already excludes them too, so this is only load-bearing for a bare `number`.)
    if ((typeNode.primitive === 'integer' || typeNode.primitive === 'number') && !Number.isFinite(value)) {
      throw new ValidationError(`${path}: expected a finite ${typeNode.primitive}, got ${value}`)
    }
    // `number` admits any JSON number; `integer` rejects a fractional value
    // (5.7) while still accepting one written 5.0 (Number.isInteger(5.0) is true).
    if (typeNode.primitive === 'integer' && !Number.isInteger(value)) {
      throw new ValidationError(`${path}: expected an integer, got ${value}`)
    }
    // An inline literal choice (project_bidi_schema.mjs's enumNode()) carries both
    // `primitive` and `enum` — the primitive check above narrows the type, but the
    // closed vocabulary below still needs to run, so only return early when there
    // is no `enum` to fall through to.
    if (typeNode.enum === undefined) return value
  }

  if (typeNode.const !== undefined) {
    if (value !== typeNode.const) {
      throw new ValidationError(
        `${path}: expected constant ${JSON.stringify(typeNode.const)}, got ${JSON.stringify(value)}`,
      )
    }
    return value
  }

  if (typeNode.enum !== undefined) {
    if (!typeNode.enum.includes(value)) {
      throw new ValidationError(
        `${path}: "${value}" is not a valid value; expected one of: ${typeNode.enum.join(', ')}`,
      )
    }
    return value
  }

  if (typeNode.list !== undefined) {
    if (!Array.isArray(value)) {
      throw new ValidationError(`${path}: expected a list, got ${typeof value}`)
    }
    return Object.freeze(value.map((item, i) => validateValue(typeNode.list, item, `${path}[${i}]`, direction)))
  }

  if (typeNode.map !== undefined) {
    if (typeof value !== 'object' || Array.isArray(value) || value === null) {
      throw new ValidationError(`${path}: expected an object, got ${typeof value}`)
    }
    // Object.create(null), not `{}`: `key` is wire-controlled and a literal "__proto__"
    // entry assigned via bracket notation would hijack result's prototype instead of
    // becoming a data property (CWE-1321) — a null-prototype object has no such trap.
    const result = Object.create(null)
    for (const [key, entry] of Object.entries(value)) {
      result[key] = validateValue(typeNode.map, entry, `${path}.${key}`, direction)
    }
    return Object.freeze(result)
  }

  if (typeNode.ref !== undefined) {
    const referenced = resolve(typeNode.ref)
    if (referenced === undefined) return value // not yet registered — best-effort, skip deep validation

    if (referenced.kind === 'enum') {
      if (!referenced.includes(value)) {
        throw new ValidationError(
          `${path}: "${value}" is not a valid ${typeNode.ref} value; expected one of: ${referenced.values.join(', ')}`,
        )
      }
      return value
    }

    if (referenced.kind === 'record') {
      if (value instanceof referenced.RecordClass) return value // already validated
      if (typeof value !== 'object' || Array.isArray(value) || value === null) {
        throw new ValidationError(`${path}: expected an object, got ${typeof value}`)
      }
      // Recurse through the same-direction path so a nested field gets the same
      // tolerance (inbound) or strictness (outbound) as its parent. Inbound returns
      // the parsed instance itself, so a typed parent record ends up with a typed
      // nested value instead of the raw wire object; outbound only validates this
      // way (the caller's own value is what gets kept, see the note above).
      if (direction === 'inbound') {
        return referenced.RecordClass.fromWire(value)
      }
      new referenced.RecordClass(value)
      return value
    }

    if (referenced.kind === 'union') {
      if (direction === 'inbound') {
        return referenced.fromWire(value)
      }
      referenced.build(value)
      return value
    }

    if (referenced.kind === 'alias') {
      return validateValue(referenced.type, value, path, direction)
    }

    return value
  }

  if (typeNode.union !== undefined) {
    const errors = []
    for (const variant of typeNode.union) {
      try {
        return validateValue(variant, value, path, direction)
      } catch (err) {
        errors.push(err.message)
      }
    }
    throw new ValidationError(`${path}: value did not match any variant (${errors.join('; ')})`)
  }

  // An inline (unnamed) record type node — project_bidi_schema.mjs's projectEntry()
  // emits this for an anonymous CDDL group (e.g. a field typed as an inline `{ ... }`
  // rather than a ref to a named, defineRecord()'d type) — same FieldSpec shape as a
  // named record's `fields`, just with nowhere to register a class. Gets the same
  // directional required/extra/nested-value handling a named record's constructor/
  // fromWire gives its fields, just built inline instead of through a Record class.
  if (typeNode.record !== undefined) {
    if (typeof value !== 'object' || Array.isArray(value) || value === null) {
      throw new ValidationError(`${path}: expected an object, got ${typeof value}`)
    }
    const byWire = new Map(typeNode.record.map((f) => [f.wire, f]))
    const result = {}
    for (const field of typeNode.record) {
      if (!Object.hasOwn(value, field.wire)) {
        if (field.required) {
          throw new ValidationError(`${path}.${field.wire}: required field is missing`)
        }
        continue
      }
      result[field.name] = validateValue(field.type, value[field.wire], `${path}.${field.wire}`, direction)
    }
    for (const wireKey of Object.keys(value)) {
      if (byWire.has(wireKey)) continue
      // No `extensible` concept exists for an inline record (project_bidi_schema.mjs
      // never sets it there) — undeclared keys get exactly the non-extensible named-
      // record treatment: rejected outbound, dropped-with-a-warning inbound.
      if (direction === 'inbound') {
        process.emitWarning(`${path}: undeclared property "${wireKey}"`, 'BiDiSchemaWarning')
      } else {
        throw new ValidationError(`${path}: unknown property "${wireKey}"`)
      }
    }
    return Object.freeze(result)
  }

  return value
}

/**
 * Registers a schema `record` — a fixed set of named fields, each independently
 * validated on the way out (constructor) and in (fromWire()).
 * @param {string} name Schema type name, e.g. 'network.AddInterceptParameters'.
 * @param {Array<{name: string, wire: string, required: boolean, type: object}>} fields
 * @param {{extensible?: boolean}} [options]
 * @returns {{new (data: object): object, fromWire: function(unknown): object}}
 *   The generated Record class — `new Record(data)` validates and constructs
 *   outbound, `Record.fromWire(payload)` validates and parses inbound.
 */
function defineRecord(name, fields, options = {}) {
  const { extensible = false } = options
  const byWire = new Map(fields.map((f) => [f.wire, f]))
  // JS property name -> wire key, the inverse of byWire — lets toJSON() below map an
  // outbound instance's own (JS-facing) properties back to the wire's declared names.
  const byName = new Map(fields.map((f) => [f.name, f.wire]))

  class Record {
    // Outbound: strict. Any value that doesn't match its declared shape is an error here.
    constructor(data) {
      if (typeof data !== 'object' || data === null || Array.isArray(data)) {
        throw new ValidationError(`${name}: expected an object`)
      }

      for (const field of fields) {
        if (!Object.hasOwn(data, field.wire)) {
          if (field.required) {
            throw new ValidationError(`${name}.${field.wire}: required field is missing`)
          }
          continue
        }
        const value = data[field.wire]
        this[field.name] = validateValue(field.type, value, `${name}.${field.wire}`, 'outbound')
      }

      for (const wireKey of Object.keys(data)) {
        if (byWire.has(wireKey)) continue
        if (!extensible) {
          throw new ValidationError(`${name}: unknown property "${wireKey}"`)
        }
        // Object.defineProperty, not `this[wireKey] = ...`: wireKey is caller-supplied
        // and a literal "__proto__" key assigned via bracket notation hijacks this
        // instance's actual prototype instead of becoming a field (CWE-1321).
        // defineProperty always creates a genuine own data property, regardless of name.
        Object.defineProperty(this, wireKey, {
          value: data[wireKey], // vendor extras reach the wire on an extensible type
          enumerable: true,
          writable: true,
          configurable: true,
        })
      }

      Object.freeze(this)
    }

    // Inbound: tolerant of undeclared properties, but a missing required field
    // is rejected just like a structurally invalid value — omission used to be
    // tolerated here, but that's no longer required.
    // Bypasses the constructor above entirely — a single constructor enforcing
    // both directions symmetrically would make tolerated undeclared-property
    // retention impossible.
    static fromWire(payload) {
      if (typeof payload !== 'object' || payload === null || Array.isArray(payload)) {
        throw new ValidationError(`${name}: expected an object on the wire, got ${typeof payload}`)
      }

      const instance = Object.create(Record.prototype)

      for (const field of fields) {
        if (!Object.hasOwn(payload, field.wire)) {
          if (field.required) {
            throw new ValidationError(`${name}.${field.wire}: required field is missing`)
          }
          continue // left genuinely absent, optional field
        }
        const value = payload[field.wire]
        // A present value's shape is never tolerated, inbound or outbound.
        instance[field.name] = validateValue(field.type, value, `${name}.${field.wire}`, 'inbound')
      }

      for (const wireKey of Object.keys(payload)) {
        if (byWire.has(wireKey)) continue
        // An extensible type preserves an undeclared field silently — no narrower
        // criterion than "extensible" itself (not, say, only fields that happen to be
        // sendable back on some other type). A non-extensible type warns and drops it
        // instead — the warning belongs only to the drop, not the retention.
        if (extensible) {
          // Object.defineProperty, not `instance[wireKey] = ...` — see the matching
          // comment in the constructor above: a literal "__proto__" key from an
          // untrusted wire payload must become a field, not swap the prototype (CWE-1321).
          Object.defineProperty(instance, wireKey, {
            value: payload[wireKey],
            enumerable: true,
            writable: true,
            configurable: true,
          })
        } else {
          process.emitWarning(`${name}: undeclared property "${wireKey}"`, 'BiDiSchemaWarning')
        }
      }

      Object.freeze(instance)
      return instance
    }

    // The JSON.stringify hook: an instance stores its fields under their JS-facing
    // property names (this[field.name]), but the wire needs the spec's own key
    // (field.wire) — the two differ whenever a generator picks an idiomatic JS name
    // distinct from the raw spec key. Runs automatically wherever this instance is
    // serialized (directly, or nested inside another value being stringified), so a
    // caller never has to remember to call it.
    toJSON() {
      // Object.create(null), not `{}`: an extra's key is wire-controlled (extensible
      // types preserve undeclared properties verbatim, see the constructor above), and
      // a literal "__proto__" key assigned via bracket notation would hijack wire's
      // prototype instead of becoming a data property (CWE-1321) — same hazard the
      // constructor/fromWire already guard against for the instance itself.
      const wire = Object.create(null)
      for (const key of Object.keys(this)) {
        wire[byName.get(key) ?? key] = this[key] // extras have no JS-name mapping — already wire-keyed
      }
      return wire
    }
  }

  Object.defineProperty(Record, 'name', { value: name })
  register(name, { kind: 'record', RecordClass: Record })
  return Record
}

/**
 * Registers a schema `alias` — a name with no fields of its own, just a
 * pointer to another type node (e.g. `network.Intercept` aliasing a plain
 * string). A ref to an alias validates through the aliased type node.
 * @param {string} name
 * @param {object} type The schema's `type` node this name aliases.
 */
function defineAlias(name, type) {
  register(name, { kind: 'alias', type })
}

module.exports = { defineRecord, defineAlias, ValidationError }
